src/HomogeneousCalorimeter_geo.cpp

```diff
@@ -406,58 +406,114 @@ static std::tuple<int, int> add_disk(Detector& desc, Assembly& env, xml::Collect
+  Volume ring12_vol("ring12", solid_ring12, outer_ring_material);
+  Transform3D tr_global_Oring = RotationZYX(15.*degree, 0., 0.) * Translation3D(0., 0., -20.*cm);
   ring12_vol.setVisAttributes(desc.visAttributes(plm.attr<std::string>(_Unicode(vis_struc))));
 
+
+
+
+  //=============================
   // Inner supporting frame
+  //=============================
+
+  // // Version1: circle shape
+  // //
+  // Tube        Ssolid_ring12(8.5*cm, rmin, calo_module_length/2., phimin, phimax);
+  // Volume      Sring12_vol("Sring12", Ssolid_ring12, inner_ring_material);
+  // Sring12_vol.setVisAttributes(desc.visAttributes(plm.attr<std::string>(_Unicode(vis_struc))));
+  // Transform3D tr_global_Iring = RotationZYX(0., 0., 0.) * Translation3D(0., 0., 0.);
+
+
+  // // Version2: elliptical shape
+  // //
+  // EllipticalTube   inner_elliptical_support_a(9.*cm, 7.5*cm, calo_module_length/2.);
+  // EllipticalTube   inner_elliptical_support_b(8.5*cm, 7.*cm, calo_module_length/2.);
+  // SubtractionSolid inner_elliptical_support_substract(inner_elliptical_support_a, inner_elliptical_support_b, Position(0., 0., 0.));
+  // Volume           inner_elliptical_vol("inner_elliptical", inner_elliptical_support_substract, inner_ring_material);
+  // inner_elliptical_vol.setVisAttributes(desc.visAttributes(plm.attr<std::string>(_Unicode(vis_struc))));
+  // Transform3D tr_global_Iring_elli = RotationZYX(-15.*degree, 0., 0.) * Translation3D(2.05*cm, 0., 0.);
+
+
+  // Version3: solid with elliptical inside
   //
-  Tube   Ssolid_ring12(8.5 * cm, rmin, calo_module_length / 2., phimin, phimax);
-  Volume Sring12_vol("Sring12", Ssolid_ring12, ring_material);
-  Sring12_vol.setVisAttributes(desc.visAttributes(plm.attr<std::string>(_Unicode(vis_struc))));
-  Transform3D tr_global_Iring = RotationZYX(0., 0., 0.) * Translation3D(0., 0., 0.);
-
-  // Supporting frame for the cabling
-  // Having overlapped with tracker barrelendcapsupporting structure, comment here until the size and position
-  // determined
-  //
+  Box inner_support_main(8.2*cm, 6.15*cm, calo_module_length/2.);  // Original size
+  // Box inner_support_main(8.2*cm, 5.125*cm, calo_module_length/2.);  // Adapted size
+  Box subtract_corner(1.025*cm, 1.025*cm, calo_module_length/2.);
+  EllipticalTube   subtract_a(7.5*cm, 5.5*cm, calo_module_length/2.);
+  SubtractionSolid inner_support_substractb1(inner_support_main, subtract_corner, Position(7.175*cm, 5.125*cm, 0.));
+  SubtractionSolid inner_support_substractb2(inner_support_substractb1, subtract_corner, Position(7.175*cm, -5.125*cm, 0.));
+  SubtractionSolid inner_support_substractb3(inner_support_substractb2, subtract_corner, Position(-7.175*cm, 5.125*cm, 0.));
+  SubtractionSolid inner_support_substractb4(inner_support_substractb3, subtract_corner, Position(-7.175*cm, -5.125*cm, 0.));
+  SubtractionSolid inner_support_substracta(inner_support_substractb4, subtract_a, Position(0., 0., 0.));
```

Comment:
Avoid serial solid subtractions because they are slow, in particular in high-occupancy regions. Use ExtrudedPolygon instead.

Reply:
Thanks for the advice, I think it will be more elegant to do this. Do you have any example can help me how to implement  ExtrudedPolygon? I tried to search this, but they are not clear to me.

Reply:
remove a serial of solid subtractions and use the ExtrudedPolygon instead.